src/MatrixFields/matrix_multiplication.jl

```diff
@@ -0,0 +1,381 @@
+    ld1 ≤ prod_d - ld2,
+    prod_d - ud2 ≤ ud1, and
+    prod_d - ud2 ≤ prod_d - ld2.
+By definition, ld1 ≤ ud1 and ld2 ≤ ud2, so the first and last inequality are
+always true. Rearranging the remaining two inequalities gives us
+    ld1 + ld2 ≤ prod_d ≤ ud1 + ud2.
+=#
+
+struct TopLeftMatrixCorner <: Operators.AbstractBoundaryCondition end
+struct BottomRightMatrixCorner <: Operators.AbstractBoundaryCondition end
+
+Operators.has_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::Operators.LeftBoundaryWindow{name},
+) where {name} = true
+Operators.has_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::Operators.RightBoundaryWindow{name},
+) where {name} = true
+
+Operators.get_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::Operators.LeftBoundaryWindow{name},
+) where {name} = TopLeftMatrixCorner()
+Operators.get_boundary(
+    ::MultiplyColumnwiseBandMatrixField,
+    ::Operators.RightBoundaryWindow{name},
+) where {name} = BottomRightMatrixCorner()
+
+Operators.stencil_interior_width(
+    ::MultiplyColumnwiseBandMatrixField,
+    matrix1,
+    arg,
+) = ((0, 0), outer_diagonals(eltype(matrix1)))
+
+# Interior indices of a center-to-center or face-to-face matrix.
+matrix_left_interior_index(space, lbw::Integer) =
+    Operators.left_idx(space) - lbw
+matrix_right_interior_index(space, rbw::Integer) =
+    Operators.right_idx(space) - rbw
+
+# Interior indices of a face-to-center matrix.
+matrix_left_interior_index(
+    space::Union{
+        Spaces.CenterFiniteDifferenceSpace,
+        Spaces.CenterExtrudedFiniteDifferenceSpace,
+    },
+    lbw::PlusHalf,
+) = Operators.left_idx(space) - lbw - half
+matrix_right_interior_index(
+    space::Union{
+        Spaces.CenterFiniteDifferenceSpace,
+        Spaces.CenterExtrudedFiniteDifferenceSpace,
+    },
+    rbw::PlusHalf,
+) = Operators.right_idx(space) - rbw + half
+
+# Interior indices of a center-to-face matrix.
+matrix_left_interior_index(
+    space::Union{
+        Spaces.FaceFiniteDifferenceSpace,
+        Spaces.FaceExtrudedFiniteDifferenceSpace,
+    },
+    lbw::PlusHalf,
+) = Operators.left_idx(space) - lbw + half
+matrix_right_interior_index(
+    space::Union{
+        Spaces.FaceFiniteDifferenceSpace,
+        Spaces.FaceExtrudedFiniteDifferenceSpace,
+    },
+    rbw::PlusHalf,
+) = Operators.right_idx(space) - rbw - half
+
+Operators.left_interior_idx(
+    space::Spaces.AbstractSpace,
+    ::MultiplyColumnwiseBandMatrixField,
+    ::TopLeftMatrixCorner,
+    matrix1,
+    arg,
+) = matrix_left_interior_index(space, outer_diagonals(eltype(matrix1))[1])
+Operators.right_interior_idx(
+    space::Spaces.AbstractSpace,
+    ::MultiplyColumnwiseBandMatrixField,
+    ::BottomRightMatrixCorner,
+    matrix1,
+    arg,
+) = matrix_right_interior_index(space, outer_diagonals(eltype(matrix1))[2])
+
+function rmul_type(::Type{T1}, ::Type{T2}, ::Type{LG}) where {T1, T2, LG}
+    type = Base._return_type(rmul_with_projection, Tuple{T1, T2, LG})
```

Comment:
I'm not so keen on reaching into internals like this. Can we work around it somehow?